library/aesce.c

```diff
@@ -0,0 +1,343 @@
+{
+    uint32_t *rki;
+    uint32_t *rko;
+    uint32_t *rk_u32 = (uint32_t *) rk;
+    memcpy(rk, key, (192 / 8));
+
+    for (size_t i = 0; i < 8; i++) {
+        rki = rk_u32 + i * (192 / 32);
+        rko = rki + (192 / 32);
+        rko[0] = ror32_8(aes_sub(rki[(192 / 32) - 1])) ^ rcon[i] ^ rki[0];
+        rko[1] = rko[0] ^ rki[1];
+        rko[2] = rko[1] ^ rki[2];
+        rko[3] = rko[2] ^ rki[3];
+        if (i < 7) {
+            rko[4] = rko[3] ^ rki[4];
+            rko[5] = rko[4] ^ rki[5];
+        }
+    }
+}
+
+/*
+ * Key expansion, 256-bit case
+ */
+TARGET_ATTR
+static void aesce_setkey_enc_256(unsigned char *rk,
+                                 const unsigned char *key)
+{
+    uint32_t *rki;
+    uint32_t *rko;
+    uint32_t *rk_u32 = (uint32_t *) rk;
+    memcpy(rk, key, (256 / 8));
+
+    for (size_t i = 0; i < 7; i++) {
+        rki = rk_u32 + i * (256 / 32);
+        rko = rki + (256 / 32);
+        rko[0] = ror32_8(aes_sub(rki[(256 / 32) - 1])) ^ rcon[i] ^ rki[0];
+        rko[1] = rko[0] ^ rki[1];
+        rko[2] = rko[1] ^ rki[2];
+        rko[3] = rko[2] ^ rki[3];
+        if (i < 6) {
+            rko[4] = aes_sub(rko[3]) ^ rki[4];
+            rko[5] = rko[4] ^ rki[5];
+            rko[6] = rko[5] ^ rki[6];
+            rko[7] = rko[6] ^ rki[7];
+        }
+    }
+}
+
+/*
+ * Key expansion, wrapper
+ */
+int mbedtls_aesce_setkey_enc(unsigned char *rk,
+                             const unsigned char *key,
+                             size_t bits)
+{
+    switch (bits) {
+        case 128: aesce_setkey_enc_128(rk, key); break;
+        case 192: aesce_setkey_enc_192(rk, key); break;
+        case 256: aesce_setkey_enc_256(rk, key); break;
+        default: return MBEDTLS_ERR_AES_INVALID_KEY_LENGTH;
+    }
+
+    return 0;
+}
+
+#if defined(MBEDTLS_GCM_C)
+
+TARGET_ATTR
+static inline uint8x16_t pmull_low(uint8x16_t a, uint8x16_t b)
+{
+    return vreinterpretq_u8_p128(
+        vmull_p64(
+            (poly64_t) vget_low_p64(vreinterpretq_p64_u8(a)),
+            (poly64_t) vget_low_p64(vreinterpretq_p64_u8(b))));
+}
+
+TARGET_ATTR
+static inline uint8x16_t pmull_high(uint8x16_t a, uint8x16_t b)
+{
+    return vreinterpretq_u8_p128(
+        vmull_high_p64(vreinterpretq_p64_u8(a),
+                       vreinterpretq_p64_u8(b)));
+}
+
+TARGET_ATTR
+static inline uint8x16x2_t ghash_mult_128(uint8x16_t GH, uint8x16_t HK)
+{
+    uint8x16_t r0, r1, a, b, t0, t1;
+    uint8x16x2_t ret;
+
+    const uint8x16_t z = vdupq_n_u8(0);
+    a = GH;
+    b = HK;
+
+    r0 = pmull_low(a, b);
+    r1 = pmull_high(a, b);
+    t0 = vextq_u8(b, b, 8);
+    t1 = pmull_low(a, t0);
+    t0 = pmull_high(a, t0);
+    t0 = veorq_u8(t0, t1);
+    t1 = vextq_u8(z, t0, 8);
+    r0 = veorq_u8(r0, t1);
+    t1 = vextq_u8(t0, z, 8);
+    r1 = veorq_u8(r1, t1);
+    ret.val[0] = r0;
+    ret.val[1] = r1;
+    return ret;
+}
+
+TARGET_ATTR
+static inline uint8x16_t ghash_mult_rdc(uint8x16x2_t in)
```

Comment:
There's nothing particularly wrong here, but reducing from `uint8x16x3_t` has 2-3x better throughput as follows:
```
uint8x16x3_t ghash_mult_128(uint8x16_t a, uint8x16_t b) {
      uint8x16x3_t ret;
      uint8x16_t c = vextq_u8(b, b);
      ret.val[1] = pmull_low(a,c);
      ret.val[0] = pmull_high(a,b);
      ret.val[2] = pmull_low(a,b);
      ret.val[1] = veorq_u8(ret.val[1], pmull_low(a,c));
      return ret;
}

uint8x16_t gmul_reduce(uint8x16x3_t a) {
    uint8x16_t const Z = vdupq_n_u8(0);
    // use 'asm' as an optimisation barrier to prevent loading R from memory
    uint64x2_t r = vreinterpretq_u64_u8(vdupq_n_u8(0x87));
    asm("" : "+w"(r));
    uint8x16_t const R = vreinterpretq_u8_u64(vshrq_n_u64(r, 64 - 8));
    uint8x16_t d = a.val[0];                //     d3:d2:00:00
    uint8x16_t j = a.val[1];                //        j2:j1
    uint8x16_t g = a.val[2];                //           g1:g0
    uint8x16_t h = pmull_high(d, R);        //        h2:h1     = reduction of d3
    uint8x16_t i = pmull_low(d, R);         //           i1:i0  = reduction of d2
    uint8x16_t k = veorq_u8(j, h);          //        k2:k1     = a0*b1 + a1*b0 + h2:h1
    uint8x16_t l = pmull_high(k, R);        //           l1:l0  = reduction of k2
    uint8x16_t m = vextq_u8(Z, k, 8);       //           m1:00  = k1:00
    uint8x16_t n = veorq_u8(g, i);          //           n1:n0  = a0*b0 + reduction of d2
    uint8x16_t o = veorq_u8(n, l);          //           o1:o0
    uint8x16_t p = veorq_u8(o, m);          //           o1:o0
    return p;
}
```
There's no need to shift/combine the middle partial product to the high and low (this is also beneficial with Aggregated Reduction Method (or postponed reduction, multiplying with different powers of H)).
Fewer shifts means hugely reduced dependency chain and more parallelism (I measured about 3x better throughput with this method on M1, and 2x better on a very old Nokia Android One phone).
As a general comment, I'm glad that this library is being optimised.

Reply:
Will change it in #6918. thanks @uncleasm